gadgets/src/mul_add.rs

```diff
@@ -118,8 +134,46 @@ impl<F: Field> MulAddChip<F> {
         let col4 = meta.advice_column();
         let mut overflow = 0.expr();
 
+        let carry_lo_cols = [col0, col1, col2, col3, col4]
+            .map(|col| (col, 3))
+            .into_iter()
+            .chain([col0, col1, col2, col3].map(|col| (col, 4)))
+            .collect::<Vec<_>>();
+
+        let carry_hi_cols = [col0, col1, col2, col3, col4]
+            .map(|col| (col, 5))
+            .into_iter()
+            .chain([col0, col1, col2, col3].map(|col| (col, 6)))
+            .collect::<Vec<_>>();
+
+        let carry_lo_expr = |meta: &mut VirtualCells<F>| {
+            carry_lo_cols
+                .iter()
+                .map(|(col, rot)| meta.query_advice(*col, Rotation(*rot)))
+                .collect::<Vec<Expression<F>>>()
+        };
+
+        let carry_hi_expr = |meta: &mut VirtualCells<F>| {
+            carry_hi_cols
+                .iter()
+                .map(|(col, rot)| meta.query_advice(*col, Rotation(*rot)))
+                .collect::<Vec<Expression<F>>>()
+        };
+
+        {
+            let mut carry_cols = carry_lo_cols.clone();
+            carry_cols.append(&mut carry_hi_cols.clone());
+
+            for (col, rot) in carry_cols.into_iter() {
+                meta.lookup("mul carry range check lo/hi lookup u8", |meta| {
```

Comment:
(optimization)

This approach is very sub-optimal because it adds lookups on each cell. By reorganizing the layout to pack all limbs in dedicated columns, there could be one lookup per column instead of one per cell.

Or again, even better to use the `UIntRangeCheckChip` instead (see other comments).

Reply:
use the UIntRangeCheckChip for carry would take 5 extra advice column thus it won't reduce cost.
the u64_check took 8 (expr) rows anyway, so use u16 carry repr is an reasonable optimization.